poh-bundle-linux-x64/oh.sh

```diff
@@ -1,68 +1,196 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# https://www.gnu.org/licenses/gpl-3.0-standalone.html
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+echo ""
+echo "***************************************************"
+echo "* Warning - Do not run this script as root user ! *"
+echo "***************************************************"
+echo ""
+
+######## Environment check:
+
+######## OPENHOSPITAL Configuration
+# POH_PATH is the directory where Portable OpenHospital files are located
+# POH_PATH=/usr/local/OpenHospital
+
+# name of this shell script
+SCRIPT_NAME="oh.sh"
+
+# set current dir
+CURRENT_DIR=$PWD
+
+if [ -z $POH_PATH ]; then
+	echo "Warning - POH_PATH not found. Please set it up properly."
+		if [ ! -f ./$SCRIPT_NAME ]; then
+		echo "Error - oh.sh not found in the current PATH. Please cd the directory where POH was unzipped or set up POH_PATH properly."
+		exit 0
+		fi
+	POH_PATH=$PWD
+fi
+
+######## Software configuration - change at your own risk :-)
 
 MYSQL_DIR="mysql-5.7.30-linux-glibc2.12-x86_64"
-JAVA_DIR="jdk-11.0.7+10-jre"
+MYSQL_URL="https://downloads.mysql.com/archives/get/p/23/file"
+
+JAVA_DISTRO="OpenJDK11U-jdk_x64_linux_openj9_11.0.8_10_openj9-0.21.0"
+JAVA_URL="https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.8%2B10_openj9-0.21.0"
+JAVA_DIR="jdk-11.0.8+10"
+
+MYSQL_PORT=3307
+
 OH_DIR="oh"
+DATABASE_NAME="oh"
+DATABASE_USER="isf"
+DATABASE_PASSWORD="isf123"
 DICOM_DEFAULT_SIZE="4M"
 
-where_i_am=$(pwd)
-cd $(dirname $0)
-POH_PATH=$(pwd)
-cd $where_i_am
+cd $POH_PATH
 
-# Find a free port to run MySQL starting from the default port.
-mysql_port=3306
-while [ $(ss -tna | awk '{ print $4 }' | grep ":$mysql_port") ]; do
-	mysql_port=$(expr $mysql_port + 1)
-done
+######################## DO NOT EDIT BELOW THIS LINE ########################
+
+echo "Checking for software...."
+
+if [ ! -d "$POH_PATH/$MYSQL_DIR" ]; then
+
+		if [ ! -f "$POH_PATH/$MYSQL_DIR.tar.gz" ]; then
+
+		echo "Warning - MySQL  not found. Do you want to download it ? (-630 Mb)"
+
+		read -p "(y/n)?" choice
+		case "$choice" in 
+			y|Y ) echo "yes";;
+			n|N ) echo "Exiting..."; exit 0;;
+			* ) echo "Invalid";;
+		esac
+
+		# Downloading mysql binary
+		echo "Downloading $MYSQL_DIR..."
+		wget $MYSQL_URL/$MYSQL_DIR.tar.gz
+		echo "Unpacking $MYSQL_DIR..."
+		tar zxvf $MYSQL_DIR.tar.gz
+		echo "Done !"
```

Comment:
this is good, but on the other hand I don't see a compelling use case for it.
I mean: if we distribute this script with the Portable, and if the Portable contains MySQL and JRE, then unless someone tampers with the distribution these lines are never executed.
Also, I think one of the goals of the Portable is being able to execute the whole application in zones without internet connection, so it would be impractical to download MySQL/JRE there.

Reply:
Everything is optional here, but there are many cases where this can be useful.
Let's say as example that I want to try a different JDK, or that I want to do a quick / automatic deploy on different environment, or let's say I want to automate stuff in a network environment.
From a SYSADMIN point of view this can be useful - and it doesn't change at all the standard behaviour of the script.

Moreover, I agree with your issue about distributing a mysql binary INSIDE a POH distro and this is something I absolutely don't like. So at least we have the option not to include it....

Reply:
Yes, as @pviotti says we should consider the context and the real use-cases... although this way the script looks more controllable it has the downside that we have to keep aligned in some way MYSQL_URL and JAVA_URL variables between oh.sh and build_poh.sh (before we had MYSQL_DIR and JAVA_DIR anyway).

Also, I think sysadmins and devs they should already have their environment to do such tests and they would not start from a portable release; instead here the suggested workflow should be the following?

1. delete mysql and java folders from the unzipped release
2. change MYSQL_URL, JAVA_URL, MYSQL_DIR, JAVA_DIR to the desidered values (e.g. MySQL 8 and Java14)
3. re-run the script and accept the download (Y)

Reply:
I think that this download feature:
- doesn't change in any way the "normal" behaviour of the script with the complete distribution
- it adds the option to distribute only a minimal version of the package
- it adds the option of updating quickly java and mysql version just by changing variables, without deleting the existing JRE and mysql bundled
- it "solves" the possible issue of packaging external software like mysql
- it's a basic feature that might start to add some simple user interactive option, which we might introduce also for the --init function
- and from a sysadm point of view, a small script that does the download it's always a "nice2have" :-)
I would definitely keep it if doesn't hurt :-)